### ISSUE TITLE
[RFC] Rethink our BS 3/4 build system

### ISSUE THREAD

**Issue by simonihmig** [kaliber5/ember-bootstrap#984]
## Motivation 

Based on [discussions](https://github.com/kaliber5/ember-bootstrap/pull/981#pullrequestreview-351928655) related to refactoring component to outerHTML semantics - as a stepping stone to later do a full migration to Glimmer components - it became clear that outerHTML semantics cause even more code duplication as before. We now have to have many more separate BS3/4 templates, as the elements root element cannot be tweaked anymore by the backing class (no more `@classNames`, `@classNameBindings` decorators), for slight difference in class names for example.

Also the often useless, but for our current build system required [re-exports by bs3/4 classes](https://github.com/kaliber5/ember-bootstrap/blob/master/addon/components/bs4/bs-tooltip.js) are annoying. 

## Detailed Design

We will introduce macro helpers, that allow us to create (JS/Template) code that differs based on the configured Bootstrap version. These helpers would look like (and initially be) regular helpers, but provide the possibility to be compiled away at build time, so to not have any runtime performance/bundle size overhead. This approach is inspired by established patterns in the Ember ecosystem, like:
* `@ember/debug` / `@glimmer/env`
* [`ember-composability-helpers`](https://github.com/pzuraq/ember-compatibility-helpers)
* [`embroider` macros](https://github.com/emberjs/rfcs/blob/embroider/text/0000-embroider-v2-package-format.md#macro-system)

### JS

A `hasBootstrapVersion` util function is introduced:

```
import { hasBootstrapVersion } from 'ember-bootstrap/compatibility-helpers';

if (hasBootstrapVersion(3)) {
  // do something BS3 specific
}
```

By a custom Babel plugin, this could be compiled down to this when using BS3:

```
if (true) {
  // do something BS3 specific
}
```

For BS4:

```
if (false) {
  // do something BS3 specific
}
```

This would later be eliminated completely by `terser` as part of Ember's minification process for production builds, as this will be recognized as deed code. Just as `@ember/debug` stuff works.

### Handlebars

A `hasBootstrapVersion` helper is introduced:

```hbs
<div class="alert" {{if this.showAlert (if (has-bootstrap-version 3) "in" "show")}}">
```

With a custom AST transform, this could be compiled down to this when using BS3:

```hbs
<div class="alert" {{if this.showAlert "in"}}">
```

### Broccoli tree magic

Our current system of having `common`/ `base` / `bs3` / `bs4` trees could be extended for a transition period, and later eventually removed completely.

By using the above helpers, the need to have BS3/Bs4 specific classes/templates is reduced. A lot of the templates will be moved to the `common` tree.

For classes, we can introduce an equivalent `common` tree, that get's merged with the `bs3`/ `bs4` tree at build time. So all components that don't need BS specific code (except for the util introduced above) can be moved to `common`, eliminating the need for re-exports from `base`.

This allows to migrate components step by step. Eventually (hopefully) all code (js/templates) will land in the `common` tree. This would allow us to completely remove the existing tree magic including the `common` parent folder and treat the addon code just like any other addon. Also the move to build-time helpers and away from broccoli tree magic seems to align well with `embroider`'s philosophy, easing the eventual migration to a native `embroider` addon when replacing our custom "macros" with those of `@ember/macros`.

## Transition period

1. Wait for the current outerHTML migrations to settle, to not cause unnecessary churn / merge conflicts
2. provide the helpers with a basic runtime implementation
3. add the `common` tree build changes (see above)
4. start refactoring existing code to use the macro helpers, moving files to `common`
5. refactor the helpers to build time AST transforms
6. remove the broccoli tree magic when all component code is in `common`

## Open questions

* When should we plan do this transition? Before cutting a v4 release, or step by step during the v4.x period?

**Comment by jelhan** [comment]
Thanks a lot for this awesome RFC. I think this is the right direction to move.

> Also the move to build-time helpers and away from broccoli tree magic seems to align well with `embroider`'s philosophy, easing the eventual migration to a native `embroider` addon when replacing our custom "macros" with those of `@ember/macros`.

We might be able to build our custom macros on top of the existing [`@embroider/macros`](https://github.com/embroider-build/embroider/tree/master/packages/macros) right now. They claim to be compatible with both Ember CLI and Embroider builds:

> This package works in both Embroider builds and normal ember-cli builds, so that addon authors can switch to this newer pattern without disruption.

Their API is a little bit different from what is proposed in [v2 Addon Format (Embroider compatibilty) RFC](https://github.com/emberjs/rfcs/pull/507). The most importance difference for our use case that I have noticed so far is the handlebars condition handling. `@embroider/macros` provide a `{{macroIf}}` helper while the RFC proposes a `{{macroCondition}}` helper that should be used together with native `{{if}}`:

```hbs
{{! @embroider/macros }}
{{#macroIf (macroGetOwnConfig "shouldUseThing") }}
  <Thing />
{{else}}
  <OtherThing />
{{/macroIf}}
```

```
{{! RFC }}
{{#if (macroCondition (macroGetOwnConfig "shouldUseThing")) }}
  <Thing />
{{else}}
  <OtherThing />
{{/if}}
```

I don't think this would be a blocker for us. If I got it right, you are suggesting to use domain-specific macros (`bootstrapVersion`) in our code. So only the AST transform of that macros would be affected by this syntax change.

> When should we plan do this transition? Before cutting a v4 release, or step by step during the v4.x period?

I don't see a risk of breaking changes here. It might be useful to enable us to do it sooner than later cause it may speed up development. But I don't think this should be a blocker for `v4` release.

> Motivation

While I agree with all points listed in this chapter, it's missing an important one in my opinion. Let me cite the Embroider RFC:

> While we guarantee that branch elimination will run in production builds, we don't guarantee that in development. The macros are designed so that in development they may have runtime implementations. This is powerful because it lets us produce a single build that works in multiple contexts.

This could be very useful for us for two scenarios:

- We could run the tests for Bootstrap 3 and Bootstrap 4 against the same build, which reduces the testing costs.
- We could switch between the Bootstrap versions at run-time, which might be a very helpful feature for the docs.

**Comment by simonihmig** [comment]
> We might be able to build our custom macros on top of the existing @embroider/macros

That seems like a very nice approach!

> Their API is a little bit different from what is proposed in v2 Addon Format (Embroider compatibilty) RFC.

Yes, but it seems the RFC API is the more recent, and the existing macros will be updated: https://github.com/embroider-build/embroider/pull/344

That would be even better for us, as we don't need that special `macroIf`, so we can indeed create an AST transform that translates our domain specific `has-bootstrap-version` to those macros pretty easily I hope. 

Or we drop our own helpers and just use the embroider ones, after all it's just a private implementation detail. It does make the code a bit more verbose though...

**Comment by jelhan** [comment]
> Or we drop our own helpers and just use the embroider ones, after all it's just a private implementation detail. It does make the code a bit more verbose though...

Maybe we do a combination of both? We keep a custom helper `hasBootstrapVersion` and use that one together with `macroCondition`?

```hbs
{{#if (macroCondition (hasBootstrapVersion 3)) }}
  <Thing />
{{else}}
  <OtherThing />
{{/if}}
```

I'm not sure if we would even need a macro condition in that case or this would be supported out of the box as long as the truth status of `hasBootstrapVersion` could be determined statically.

**Comment by jelhan** [comment]
The Embroider RFC includes a paragraph why they propose an explicit `macroCondition` function and helper:

> Q: Why not allow if (getOwnConfig().thing) instead of if (macroCondition(getOwnConfig().thing))?
> 
> A: Because we don't want to leave any confusion over whether branch elimination will be done. Boolean expressions that include a macro like getOwnConfig alongside other runtime-only values are perfectly legal. But those expressions would not allow branch elimination. The ambiguity means you might accidentally defeat branch elimination without noticing. macroCondition is intended to signal -- both to the reader and to the compiler -- that this place absolutely must do branch elimination. It's an error if we can't eliminate one branch or the other.

That's a strong argument that we should consider carefully in my opinion.

**Comment by simonihmig** [comment]
@jelhan picking up your suggestion [here](https://github.com/kaliber5/ember-bootstrap/pull/996#discussion_r377868451):
I see your point, and indeed I also realized that the way this is written is not so favorable for eventually introducing BS5 flavors of the components. Your suggestion is quite verbose though...

What about introducing a new helper, somewhat like a switch statement (vs. long chains of "ifs"):

```hbs
class="... {{if this.showContent (switch-bootstrap 4="show" 3="in")}}"
```

This could as well be transformed by an AST transform to a static value.

**Comment by jelhan** [comment]
> What about introducing a new helper, somewhat like a switch statement (vs. long chains of "ifs"):
> 
> ```
> class="... {{if this.showContent (switch-bootstrap 4="show" 3="in")}}"
> ```
> 
> This could as well be transformed by an AST transform to a static value.

I see the point that the solution I had proposed here is very verbose. But I'm not convinced by the switch based approach to be honest:

- It does not support block-mode. Or did I missed something?
- How to express if a class is not needed for all Bootstrap versions? Would we need an additional `{{has-bootstrap-version}}` helper? Or must `{{switch-bootstrap}}` support branch elimination if it does not have a value for used bootstrap version?
- It couldn't be used together with `macroCondition` cause it does not return a boolean value. As said in my last comment in this issue I see strong arguments to also use `macroCondition` in our version-specific build system.
- It would be very different from established patterns in the ember community. I fear we would just replace one entrance barrier for new contributors by another.
- A named argument that is only a number looks uncommon to me. Actually my first reaction was: Is this even valid syntax for glimmer templates?

I'm leaning more towards a solution that is directly based on the macro system proposed in Embroider RFC. We may could rewrite the [example discussed](https://github.com/kaliber5/ember-bootstrap/pull/996#discussion_r377868451) like this:

```hbs
class={{bs-join
  (if this.collapse "collapse")
  (if this.collapsing "collapsing")
  (if (macroCondition (has-bootstrap-version 3)) (if this.showContent "in"))
  (if (macroCondition (has-bootstrap-version 4)) (if this.showContent "show"))
}}
```

This assumes that other template helpers could be used together with `macroCondition` as long as the value could be determined statically. I'm not sure if that assumption is correct. I raised the question here: https://github.com/emberjs/rfcs/pull/507/files#r379162904 But even if not an AST transform could be used to support such an authoring format.

I also introduce a `{{bs-join}}` helper in that snippet. It should help to format the code nicely without introducing unnecessary whitespaces. It would join all it's positional arguments as long as they are not empty with a whitespace as a separator. So something like this:

```js
import { helper } from "@ember/component/helper";
import { getConfig } from "@ember/macros"; 

function bsJoin(items) {
  return items.filter((_) => _).join(' ');
}

export default helper(bsJoin);
```

I also think that the order in nested `{{if}}` is critical as soon as it comes to branch elimination. In that case the branch elimination must be done at the most outer `{{if}}` cause otherwise parts of the statement would remain. E.g. `{{if this.showContent (if (has-bootstrap-version 3) "in")}}` would be after branch elimination `{{if this.showContent undefined}}`. Or did I missed something? This is important as soon as we deal with scenarios in which a conditional class is only required for a specific bootstrap version but not at all for the other.

**Comment by simonihmig** [comment]
> It does not support block-mode. Or did I missed something?

Yes, but I did not suggest this as a replacement for `has-bootstrap-version`, but as syntactic sugar basically, for the *very* common use case of having two different class names.

> It couldn't be used together with macroCondition cause it does not return a boolean value. As said in my last comment in this issue I see strong arguments to also use macroCondition in our version-specific build system.

Tbh, I don't really understand why we should use `macroCondition`. In the description you quoted it was stated this is needed to make sure that it can be statically analyzed, while `getOwnConfig` allows also for runtime semantics. That's at least how I understand it. But we don't use `getOwnConfig`. So we can make `has-bootstrap-version` always fail the build if we cannot eliminate it at build time.

> It would be very different from established patterns in the ember community. I fear we would just replace one entrance barrier for new contributors by another.

Really? I mean it's a helper, just as `has-bootstrap-version`, and you do have to understand what it does in any case. But how is it more difficult to understand than any other?

> A named argument that is only a number looks uncommon to me. Actually my first reaction was: Is this even valid syntax for glimmer templates?

Yeah, I was also unsure if that actually is valid. If not we can turn it into `(switch-bootstrap bs4="show" bs3="in")`

> I also introduce a {{bs-join}} helper in that snippet. 

I see, and personally I also feel some sort of pain when seeing double or trailing whitespace in a class attribute. 😝 But I think this is orthogonal to the other discussed topics, and also not a build-time concern.  And also it could qualify as...

> It would be very different from established patterns in the ember community.

:wink: But I am open to this idea. But again, let's treat this separately for now.

**Comment by simonihmig** [comment]
But I won't insist on using that construct... 

I was just worried about the readability of the original example: 
```
{{if this.showContent (if (has-bootstrap-version 4) "show") (if (has-bootstrap-version 3) "in")}}"
```

We could also settle this discussion by committing to not use the else-part together with `has-bootstrap-version`. So basically something like you suggested above:
```
{{if (has-bootstrap-version 3) (if this.showContent "in")}}
{{if (has-bootstrap-version 4) (if this.showContent "show")}}
```

Does that make sense?

As I said I just don't really see the point of using `macroCondition`.

**Comment by jelhan** [comment]
> Tbh, I don't really understand why we should use `macroCondition`. In the description you quoted it was stated this is needed to make sure that it can be statically analyzed, while `getOwnConfig` allows also for runtime semantics. That's at least how I understand it. But we don't use `getOwnConfig`. So we can make `has-bootstrap-version` always fail the build if we cannot eliminate it at build time.

> As I said I just don't really see the point of using `macroCondition`.

For me the important sentence in the RFC about `macroCondition` is the following one:

> `macroCondition` is intended to signal -- both to the reader and to the compiler -- that this place absolutely must do branch elimination.

So using it is mostly about clarity of the code not that much about functionality.

Of course `{{has-bootstrap-version}}` or `{{switch-bootstrap}}` could also both do branch elimination. They would signal the same in our codebase if they guarantee to either do branch elimination or throw at build-time. But I fear that the signal would not be that helpful to most readers of our source code cause it would require deep knowledge about Ember Bootstrap internals as it would be an ember-bootstrap-specific solution.

I expect the macro system to become a widely used pattern in Ember addons soon after being shipped in a stable release. I assume that most experienced Ember developers will have at least a rough understanding what `macroCondition` guarantees. I think the main take away for most Ember developers will be:

- You should use `macroCondition` to strip unnecessary code at build time.
- Not using it can cause performance issues.

Assuming that these assumptions are right I fear that a custom build system which is not using `macroCondition` could cause a lot of confusion. Beside adding to the learning curve for first time contributors I'm afraid that it may lead to wrong assumptions about the code quality of Ember Bootstrap and it's build size / performance impact.

**Comment by simonihmig** [comment]
This has now been implemented by #1023, although not with custom helpers anymore, but plain `@embroider/macros`.
